### ISSUE TITLE
Automate environment initialization

### ISSUE THREAD

**Issue by matth79** [calacademy-research/COMB#32]
Knowing which R packages a script depends upon currently requires running the script repeatedly and calling packages.install() on each failure.

Some scripts also seem to rely on library() functions (e.g. fread), where the library() has not been explicitly loaded in the script.

Ideally, it would be possible, from a bare R install, to run one script:

```
Rscript install_dependencies.R
```

with a packages.install line for each library() of any script and a coding rule that requires all scripts to add explicit library() lines for every library they depend upon.

There are other alternatives. The feature request of this issue is to have a one-touch way to bootstrap a development environment on a machine that has nothing but a bare R install.

(Probably some C library dependencies are involved as well, e.g. jags.)

**Comment by mschulist** [comment]
Yes. This would also be useful when updating R since updating R requires reinstalling packages.

**Comment by ddkapan** [comment]
Hi @matth79 & @mschulist there are several options, an easy one would be to gather a unique list of all calls to library() && require() and then do the basic 'if you don't find the package installed, install it logic'.  

A better way maybe would be to create a conda environment, set up with environment.yml and load that environment before running anything.  Thoughts?

**Comment by matth79** [comment]
I did some research on this.

[renv](https://rstudio.github.io/renv/index.html) appeals to me because of its conceptual similarity to python virtualenv. It also looks like it might do what I asked in the description just for me, without needing to mandate it for all contributors. (What gave me that idea is the "dependencies are discovered" [bit](https://rstudio.github.io/renv/reference/init.html#details).)

[pacman](https://github.com/trinker/pacman/) is nice for being targeted and unmysterious. A collectively agreed-upon convention to always use pacman::p_load() rather than library() would clean up the syntax for "if not found install." However, I think people are used to using library() and that this will still clutter up the code.

I hesitate about Anaconda because of the marketing and that the selling point is R+Python, together at last, which is more than this issue requires.

I don't think any of these are a complete solution for bootstrapping (a development environment, not a [distribution](https://en.wikipedia.org/wiki/Bootstrapping_(statistics))).    ...again because of dependencies on native dynamic libraries (.dylib, .DLL, .so) that aren't usually on a fresh machine.

I'll assign this to myself for now and eventually make a couple attempts to bootstrap from standard cloud VM images. At the end of that, the artifact will probably be adding a DEVELOPING.md document to this repo with instructions, hopefully minimizing the number of steps. (I appear to assume posterity is going to be very interested in repeating what we've done here.)

**Comment by ddkapan** [comment]
Thanks, I appreciate this!

For now a hack is to have build a 'check the library' list c("package1","package2",...) to see if installed and not pass to install.package(...).

Top of each script, or overall in the first set of functions?

Or not at all?

**Comment by ddkapan** [comment]
I read the [https://cran.r-project.org/web/packages/renv/vignettes/renv.html](vignette) about renv, that sounds perfect! It will be a tiny bit of effort for us to educate ourselves, we can also help by calling renv::snapshot().  But I'm curious how this will interact with git/GitHub, thoughts? Willing to test? :) DK

**Comment by matth79** [comment]
Usually the way dependency pinners like venv, gem, and cargo interact with version control is to create some sort of "lock file" with the version numbers of the actually installed dependencies.

We very definitely do not want whatever directory(ies) end up holding all those specific versions of install libraries to be tracked in github, but typically the lock file does get submitted to keep everyone - and every server - "on the same page."

**Comment by ddkapan** [comment]
OK, so what are the next steps, and do we need to modify the .gitignore?

**Comment by matth79** [comment]
Durrell tested this and only hit one small snag.

People who pull will be prompted to run renv::restore but will also have the option not to do that if they don't want to wait ~30 minutes for packages to download and install.


As a side note:

While working on this I found that there definitely are C dependencies. On Debian Linux they are called (jags libdgal-dev libnlopt-dev libdunits2-dev libssl-dev). I used containers to test the process of bootstrapping a "devbox" to be sure exactly what the dependencies are. As a side effect, I got a "Dockerfile" and "[package](https://github.com/users/matth79/packages/container/package/comb)" that we don't really need and that I don't expect even myself to use for day-to-day work. The Dockerfile could be worth keeping in the repo for "documentary" purposes.
